### PR TITLE
Adding functionality to enable retrieving the number of evicted objec…

### PR DESCRIPTION
**Pull request by armenatzoglou**
…ts from MDCache for testing purposes

This pull request needs code that is in #145. It will be pushed to master after #145.

### CODE REVIEW

**Comment by armenatzoglou** [comment]
@foyzur @gcaragea @xinzweb Please take a look when you get a chance.

**Comment by hlinnaka** [comment]
With no callers to the new function, isn't this just dead code?

**Comment by armenatzoglou** [comment]
@hlinnaka This function is called by a user defined function in a tinc test.

**Comment by hlinnaka** [comment]
On 06/01/16 20:27, Nikos Armenatzoglou wrote:
> @hlinnaka This function is called by a user defined function in a tinc test.

Please find a way to put this with the test case, then. If it sits in 
src/backend/gpopt like this, it *will* be removed by someone a few 
months down the road, who doesn't remember why it's there.

At the very least there needs to be a comment explaining why it's not 
dead code, but that has the downside that once the test case becomes 
obsolete, this code will linger, because someone looking at this code 
will not know where the tinc test is or whether it's still needed.

- Heikki

**Comment by gcaragea** [comment]
Heikki,

We had a discussion about this yesterday in our stand-up, and my
recommendation is to go with the TINC solution for now.

The TINC test does more than just run a SQL file and compare the results.
It also compiles the C udf and loads it. AFAIK there is no easy way to do
this from ICG at this point. So rather than blocking this story
indefinitely until we find a way to port this mechanism to the ICG
framework, I'd like to keep moving and finish this feature.

Part of the big TINC --> ICG effort, we'll tackle this and other similar
tests in one go.

Thanks,
George

**Comment by armenatzoglou** [comment]
@hlinnaka I see your point. I will add a comment. Unfortunately, for the moment, the test should be in tinc, since it requires a user defined function, which is installed during test execution and accesses the value of a variable that lives in gpos world.

**Comment by hlinnaka** [comment]
On 06/01/16 21:30, George Caragea wrote:
> Heikki,
>
> We had a discussion about this yesterday in our stand-up, and my
> recommendation is to go with the TINC solution for now.
>
> The TINC test does more than just run a SQL file and compare the results.
> It also compiles the C udf and loads it.

It sounds like it would be easy to include this function in the test C 
code that it's loading anyway, then.

> AFAIK there is no easy way to do this from ICG at this point. So
> rather than blocking this story indefinitely until we find a way to
> port this mechanism to the ICG framework, I'd like to keep moving and
> finish this feature.

Where did this come from? I didn't say anything about installcheck-good.

> Part of the big TINC --> ICG effort, we'll tackle this and other similar
> tests in one go.

Well, we really should stop adding tests to TINC, but that's not what I 
was commenting on.

- Heikki

**Comment by gcaragea** [comment]
On Wed, Jan 6, 2016 at 11:35 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> > The TINC test does more than just run a SQL file and compare the results.
> > It also compiles the C udf and loads it.
>
> It sounds like it would be easy to include this function in the test C
> code that it's loading anyway, then.
>
>
I'll let Nikos comment on that, I think we might need this hook in the GPDB
side though, not sure if it can be done completely outside.



> > Part of the big TINC --> ICG effort, we'll tackle this and other similar
> > tests in one go.
>
> Well, we really should stop adding tests to TINC, but that's not what I
> was commenting on.
>


Oh, my bad, I misunderstood. I agree with you on not adding tests to TINC,
but as I said, this requires some functionality that is only in TINC now,
and we'd rather tackling that aspect more holistically.

Thanks.
George

**Comment by foyzur** [comment]
On Wed, Jan 6, 2016 at 11:22 AM, Heikki Linnakangas <
notifications@github.com> wrote:

> On 06/01/16 20:27, Nikos Armenatzoglou wrote:
> > @hlinnaka This function is called by a user defined function in a tinc
> test.
>
> Please find a way to put this with the test case, then. If it sits in
> src/backend/gpopt like this, it *will* be removed by someone a few
> months down the road, who doesn't remember why it's there.
>

Heikki do you have any suggestion about how we can incorporate this into
test code? This requires access to CMDCache, loaded from a shared library.
The corresponding headers are not (currently) exported with the binary.
Therefore, the UDF c file will not have access to the underlying APIs. My
understanding is, Nikos is only using the apparently dead code for
exporting the function in C. What he can do instead is to find a better
place to locate the C export. Otherwise we need to find a way to include
optimizer headers with the binary as TINC won't have access to headers that
are not included with the binary.

>
> At the very least there needs to be a comment explaining why it's not
> dead code, but that has the downside that once the test case becomes
> obsolete, this code will linger, because someone looking at this code
> will not know where the tinc test is or whether it's still needed.
>
> - Heikki
>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/greenplum-db/gpdb/pull/243#issuecomment-169426911>.
>



-- 

Foyzur Rahman

**Comment by hlinnaka** [comment]
On 06/01/16 22:15, foyzur wrote:
> On Wed, Jan 6, 2016 at 11:22 AM, Heikki Linnakangas <
> notifications@github.com> wrote:
>
>> On 06/01/16 20:27, Nikos Armenatzoglou wrote:
>>> @hlinnaka This function is called by a user defined function in a tinc
>> test.
>>
>> Please find a way to put this with the test case, then. If it sits in
>> src/backend/gpopt like this, it *will* be removed by someone a few
>> months down the road, who doesn't remember why it's there.
>>
>
> Heikki do you have any suggestion about how we can incorporate this into
> test code? This requires access to CMDCache, loaded from a shared library.
> The corresponding headers are not (currently) exported with the binary.
> Therefore, the UDF c file will not have access to the underlying APIs. My
> understanding is, Nikos is only using the apparently dead code for
> exporting the function in C. What he can do instead is to find a better
> place to locate the C export. Otherwise we need to find a way to include
> optimizer headers with the binary as TINC won't have access to headers that
> are not included with the binary.

Sure. Can you share the test case, please, and I'll have a look?

- Heikki

**Comment by armenatzoglou** [comment]
@hlinnaka I sent you an email regarding the location of the test. In case that we do not find a better solution, I have already moved the function to gpopt/utils/funcs.cpp, which includes an API for invoking optimizer using GPDB udfs. I think that the new location is safer. Thanks a lot for your help.

**Comment by hlinnaka** [comment]
On 07/01/16 00:03, Nikos Armenatzoglou wrote:
> @hlinnaka I sent you an email regarding the location of the test. In
> case that we do not find a better solution, I have already moved the
> function to gpopt/utils/funcs.cpp, which includes an API for invoking
> optimizer using GPDB udfs. I think that the new location is safer.

Yeah, that's better. Still missing a comment.

- Heikki

**Comment by hlinnaka** [comment]
On 06/01/16 22:15, foyzur wrote:
> Heikki do you have any suggestion about how we can incorporate this into
> test code? This requires access to CMDCache, loaded from a shared library.
> The corresponding headers are not (currently) exported with the binary.
> Therefore, the UDF c file will not have access to the underlying APIs. My
> understanding is, Nikos is only using the apparently dead code for
> exporting the function in C. What he can do instead is to find a better
> place to locate the C export. Otherwise we need to find a way to include
> optimizer headers with the binary as TINC won't have access to headers that
> are not included with the binary.

Hmm. Looking at eviction_counter.c, you're currently doing:

> extern uint64_t GetCacheEvictionCounter();

In theory, you could have an extern declaration like that, directly for 
the CMDCache::ULLGetCacheEvictionCounter() function. This seems to work, 
as a standalone program at least:

> #include <stdint.h>
> #include <stdio.h>
>
> typedef uint64_t ULLONG;
>
> namespace gpopt
> {
> class CMDCache
> {
> public:
>   static ULLONG ULLGetCacheEvictionCounter();
> };
>
>
> extern "C"
> {
> int main(int argc, char **argv)
> {
> printf("%ull", CMDCache::ULLGetCacheEvictionCounter());
> }
> }
> }

It's quite ugly, #including the right header file would certainly be 
nicer. But it's not fundamentally different from the existing extern 
declaration.

If we don't want to go down this route, and prefer to add the function 
into src/backend/gpopt, I think it would make sense to define the 
function so that you could use it directly in the CREATE FUNCTION 
statement, without needing the C code in the test suite at all. I.e. 
something like:

Datum
orca_eviction_counter(PG_FUNCTION_ARGS)
{
         PG_RETURN_INT64(CMDCache::ULLGetCacheEvictionCounter());
}

- Heikki

**Comment by armenatzoglou** [comment]
@ Heikki Thanks a lot for your help. I tried to follow your suggestions, but I faced two issues. 
Regarding your first solution: 
> you could have an extern declaration like that, directly for
the CMDCache::ULLGetCacheEvictionCounter() function.

In tinc, we use the existing gpdb makefiles (greenplum-db-devel/lib/postgresql/pgxs/src/makefiles/pgxs.mk) that allows us to have a clear and neat way of building udfs (see //tincrepo/main/mpp/common/tests/mdversion/mdversion_utils.py and //tincrepo/main/mpp/common/tests/mdversion/udfs/Makefile). Currently, to the best of my knowledge, GPDB does not support cpp compile, so I think that we should avoid this direction.

About your second suggestion:
> it would make sense to define the
> function so that you could use it directly in the CREATE FUNCTION
> statement, without needing the C code in the test suite at all  

That would be a great idea if the C code in the test suite was simply calling the CMDCache::ULLGetCacheEvictionCounter() function. However, as you saw already, it also requires to keep "states" regarding the number of eviction objects, which needs additional functionality. Also, I do not think that it is a good idea to move the C code from the test to the function in src/backend/gpopt, since we will have issues regarding the "states" that I described above. Although any test can trigger the eviction, I want only my test to be able to change the "state". Moreover, I am sure that we are going to use CMDCache::ULLGetCacheEvictionCounter() in other tests as well, so it will be good to keep the code in gpdb simple. 

To sum up, I would prefer to keep this code and the test as it is. As I mentioned in my previous comment and you agreed, the new location is safer. I will also add a comment at function's doc.

Thanks,
Nikos

**Comment by gcaragea** [comment]
Latest changes as of Jan 7 LGTM. New location and comment should prevent someone from cleaning up the UDFs by mistake.

File: src/backend/gpopt/utils/funcs.cpp [hunk 1]
```diff
@@ -1341,3 +1343,27 @@ PlannedStmt *orca(Query *pquery)
...
+//		Get the number of times we evicted entries from CMDCache.
+//		This function is called by udfs for testing purposes.
+//
+//---------------------------------------------------------------------------
+extern "C"
+{
+ULLONG
+GetCacheEvictionCounter()
+{
+	if (!CMDCache::FInitialized())
```
Comment (foyzur) [inline src/backend/gpopt/utils/funcs.cpp]: Can this ever happen during tests (i.e., an uninitialized CMDCache)? I would assume the optimizer would first generate a plan and then executor would execute the UDF. So, shouldn't we already have an initialized CMDCache? The only concern could be a fallback during planning of this query. But, I don't think we fallback for this one currently.

**Comment by foyzur** [comment]
LGTM as of the latest change that relocates to funcs.cpp.

File: src/backend/gpopt/utils/funcs.cpp (hunk 1 above)
Comment (armenatzoglou) [inline src/backend/gpopt/utils/funcs.cpp]: If the execution plan of the udf function, which uses GetCacheEvictionCounter(), is produced by planner, then CMDCache will be NULL. So, in that case, we have to initialize CMDCache. This is a realistic scenario even if optimizer is on.

**Comment by armenatzoglou** [comment]
Thanks everyone for the review. Pushed to master.
